### ISSUE TITLE
Investigate eLife-style content adapter

### ISSUE THREAD

**Issue by thewilkybarkid** [libero/libero#131]
Determine the best approach to produce an adapter that populates a Libero content API in the style that eLife currently works (which will be copied by IJM).

Known steps:

- Monitor an S3 bucket for ZIP files containing a JATS XML document and associated assets (just web-ready images for now)
- Makes its content ready for a content API
  - Wrap JATS inside a `<libero:item>`
  - Makes the asset URIs in the JATS absolute (HTTP), can use `@xml:base`
  - Makes the assets accessible at that URI
- Sends a `PUT` request to a `/items/{id}/versions/{version}` route in a content API.

## Tasks

- [x] Determine appropriate technologies to use
  - [x] Can monitor S3 bucket
  - [x] Can read ZIP content
  - [x] Can rewrite JATS XML
  - [x] Can wrap JATS XML in Libero XML
  - [x] Can send `PUT` request
  - [x] Is stateless
  - [x] Is resilient
  - [x] Has logging
- [ ] Choose a name
- [ ] Create ticket(s) for the implementation

## Notes

- Don't have to worry about versions yet, can assume everything is version '1'.
- An implement may be able monitor other places (eg Google Cloud Storage, generic FTP), but isn't a requirement.
- Doesn't need to worry about handling non-web image types (eg TIFFs); can assume everything is JPEG.
- Will need to integrate with the (currently undefined) Libero event bus in some form (emitting events as it performs tasks).

## Links

- [Content API specification](https://docs.google.com/document/d/1jMJ4NxzHiBroNerUbiNWn5H01kgaPRvfRELr1WDvPRY/edit#heading=h.qbo1gd19om1o)
- [`<libero:item>` schema](https://github.com/libero/schemas/blob/master/api/content/item.rng)
- [JATS documentation](https://jats.nlm.nih.gov/archiving/tag-library/1.2/)

## Related

- #130 will set the image markup in JATS that will be supported initially.
- #132 adds `PUT` support to the Content Store.

**Comment by giorgiosironi** [comment]
The description has good technical details already, this minimal process is the one be implemented:
1. I have a .zip package in an S3 bucket which contains a JATS XML document and N JPEG images
1. An event is added to a (SQS or else) queue
1. Something reads this message the package is downloaded
1. The package is extracted so that the files are independently accessible
1. The <article-id pub-id-type="publisher-id">19394</article-id> element is read to understand what is the article id
1. The article version is hardcoded to 1
1. The files are uploaded to a extracted (possibly temporary) bucket 
1. The <graphic mime-subtype="jpeg" mimetype="image" xlink:href="elife-19394-resp-fig1"/> references in the JATS XML are made absolute by adding a xml:base=”https://s3.amazonaws.com/some-bucket-name/elife-19394/” attribute to the top-level <article xmlns:xlink="http://www.w3.org/1999/xlink" article-type="research-article" dtd-version="1.1d3">
1. The JATS content is wrapped into a <libero> envelope
1. A PUT request is made to a content API containing a Libero-wrapped JATS with any assets references expressed as absolute HTTP(S) URIs

**Comment by GiancarloFusiello** [comment]
Continuing from @giorgiosironi comment above to further define the service operations.

Assumptions:
- Only files with extensions `.xml`, `.jpg` and `.jpeg` will be read/processed for MVP
- `.zip` files will be access from an AWS S3 bucket (What is the bucket address?)
- How `.zip` files are added to AWS S3 are not the concern of the ECA (eLife-style content adapter)

Process/processes:
1. Retrieve `.zip` file from AWS S3 and extract files (locally or write files to a new bucket/buckets?)
1. Validate xml against schema (Where can schemas be accessed from? Which schema?)
1. Add `xml:base=""` attribute to (which tag?)
1. Wrap JATS content (please define)
1. Make http PUT request to (address? is this just sending the xml document or the images also?)

@thewilkybarkid @giorgiosironi please can you edit/add missing information above?

**Comment by giorgiosironi** [comment]
Didn't want to edit comments due to race conditions/lack of notifications. My thoughts:

> Only files with extensions .xml, .jpg and .jpeg will be read/processed for MVP

We also have `.pdf` in the scope of the MVP, and possibly supplementary files like `.xslx` that could be available for download; or any generic extension which are treated opaquely without any processing. except perhaps renaming.

> .zip files will be access from an AWS S3 bucket (What is the bucket address?)

For example in the `unstable` environment it could be `unstable-elife-style-content-adapter-incoming` (prefix with an environment name, name of the application to understand its scope, plus what it's used for).

> How .zip files are added to AWS S3 are not the concern of the ECA (eLife-style content adapter)

We need to ask the production team or the content provider for explanations (I will ask).

> Retrieve .zip file from AWS S3 and extract files (locally or write files to a new bucket/buckets?)

Depends on how the work is split. If the .zip extraction and the subsequent processing of the files in it happen in a single transaction[1], files wouldn't need to be persisted externally in buckets.

However most of the files need to be accessible in a bucket for the content store to download them when handling the PUT, so they do need to be put into a bucket such as `unstable-elife-style-content-adapter-expanded` (different name).

[1] fancy name for a continuous execution inside a single container or process

> Validate xml against schema (Where can schemas be accessed from? Which schema?)

Not sure this is needed at this stage.

> Add xml:base="" attribute to (which tag?)

The root element IIRC.
 
> Wrap JATS content (please define)

This is the [JATS document](https://github.com/libero/sample-configuration/blob/master/.docker/scholarly-articles/data/article1/1.xml#L12-L152) inside the zip, and this is the [result of the wrapping](https://github.com/libero/sample-configuration/blob/master/.docker/scholarly-articles/data/article1/1.xml#L1-L154).

> Make http PUT request to (address? is this just sending the xml document or the images also?)

Only the XML, PUT is defined at https://docs.google.com/document/d/1jMJ4NxzHiBroNerUbiNWn5H01kgaPRvfRELr1WDvPRY/edit#heading=h.qbo1gd19om1o but the URL looks like:
http://unstable--api-gateway.libero.pub:8081/scholarly-articles/items/10627/versions/1
where `10627` is the id of the article and `1` is the only hardcoded version we support at the moment.

The images should be accessible through download from a bucket so they don't need to be passed in the HTTP request, which only allows one body.

**Comment by GiancarloFusiello** [comment]
Continuing from @giorgiosironi comment above to further define the service operations.

Assumptions:
- Only files with extensions `.xml`, `.pdf`, `.xslx`, `.jpg` and `.jpeg` will be read/processed for MVP
- `.zip` files will be access from an AWS S3 bucket (What is the bucket address?)
- How `.zip` files are added to AWS S3 are not the concern of the ECA (eLife-style content adapter)

Process/processes:
1. Retrieve `.zip` file from AWS S3 and extract files (locally or write files to a new bucket/buckets?)
1. Wrap JATS content with an item tag like so: 
```xml
<item xmlns="http://libero.pub" xmlns:jats="http://jats.nlm.nih.gov">
  <jats:article></jats:article>
</item>
```
1. Add `xml:base=""` attribute to document declaration tag like so:
```xml
<?xml version="1.0" encoding="UTF-8"? xml:base="http://<path-to-s3-bucket>">
```
1. (Add meta data to document?) like so:
```xml
<item xmlns="http://libero.pub" xmlns:jats="http://jats.nlm.nih.gov">
    <meta>
        <id>article1</id>
        <service>scholarly-articles</service>
    </meta>
    <jats:article></jats:article>
</item>
```
1. Make http PUT request to (address?) with the resulting xml document as the body

@thewilkybarkid @giorgiosironi please can you edit/add missing information above?

**Comment by giorgiosironi** [comment]
> 1. Retrieve .zip file from AWS S3 and extract files (locally or write files to a new bucket/buckets?)

Extract .zip files from the `unstable-elife-style-content-adapter-incoming bucket, and place the results in the `unstable-elife-style-content-adapter-expanded` bucket. The namea are fully qualified without a URL being needed, but the application will also need credentials.

There is the possibility of making these buckets private, but that will require some [additional complex authentication](https://github.com/tax/python-requests-aws) on the Libero `content-store` side that would expose it's actually S3 since it implies calculating a signature. An alternative is to generate an unpredictable prefix like `s3://unstable-elife-style-content-adapter-expanded/10627/7eb6f5e9-dc3f-4f59-91d5-7ce154b42502/elife-10627-fig1-v1.jpg` (where `7eb6f5e9-dc3f-4f59-91d5-7ce154b42502` is the generated UUID of this attempt.)

**Comment by GiancarloFusiello** [comment]
Continuing from @giorgiosironi comment above to further define the service operations.

Assumptions:
- Only files with extensions `.xml`, `.pdf`, `.xslx`, `.jpg` and `.jpeg` will be read/processed for MVP
- `.zip` files will be access from an AWS S3 bucket (What is the bucket address?)
- How `.zip` files are added to AWS S3 are not the concern of the ECA (eLife-style content adapter)

Configurations:
- Add AWS credentials
- (Add DB credentials?) 

Process/processes:
1. Retrieve `.zip` file from AWS S3 `s3://unstable-elife-style-content-adapter-incoming`
2. Extract files and add to AWS S3 `s3://unstable-elife-style-content-adapter-expanded`
3. Wrap JATS content with an item tag like so: 
```xml
<item xmlns="http://libero.pub" xmlns:jats="http://jats.nlm.nih.gov">
  <jats:article></jats:article>
</item>
```
4. Add `xml:base=""` attribute to document declaration tag like so:
```xml
<?xml version="1.0" encoding="UTF-8"? xml:base="http://<path-to-s3-bucket>">
```
5. (Add meta data to document?) like so:
```xml
<item xmlns="http://libero.pub" xmlns:jats="http://jats.nlm.nih.gov">
    <meta>
        <id>article1</id>
        <service>scholarly-articles</service>
    </meta>
    <jats:article></jats:article>
</item>
```
6. Make http PUT request to `http://unstable--api-gateway.libero.pub:8081/scholarly-articles/items/<article-id>/versions/1` with the resulting xml document as the body

@thewilkybarkid @giorgiosironi please can you edit/add missing information above?

**Comment by GiancarloFusiello** [comment]
Regarding `6.` above, could the file be retrieved from the bucket it has been expanded to rather than sent directly? I'm assuming this will need to be done for other files such as images?

**Comment by GiancarloFusiello** [comment]
Does the newly generated xml document need to be saved as a file somewhere or is sending it via a PUT request sufficient?

**Comment by giorgiosironi** [comment]
> Does the newly generated xml document need to be saved as a file somewhere or is sending it via a PUT request sufficient?

Should be the responsibility of the `content-store` to save it, so no need for storage on this side.

**Comment by giorgiosironi** [comment]
https://github.com/libero/infrastructure/pull/14 will cover AWS infrastructure including credentials. DB credentials can be avoided by firewalling the database away from the outside world, I think?

**Comment by thewilkybarkid** [comment]
> * Only files with extensions `.xml`, `.pdf`, `.xslx`, `.jpg` and `.jpeg` will be read/processed for MVP

The MVP will probably cover all file types (assuming things like supplementary material are included), only having specific handling for image types (unknown what should be supported, but will probably include converting TIFFs). For this first pass though, we're just handling JPEGs.

> * `.zip` files will be access from an AWS S3 bucket (What is the bucket address?)

Address is configurable?

> 1. Add `xml:base=""` attribute to document declaration tag like so:
> 
> ```
> <?xml version="1.0" encoding="UTF-8"? xml:base="http://<path-to-s3-bucket>">
> ```

The XML declaration isn't an element. You mean:

```
<?xml version="1.0" encoding="UTF-8"?>
<article xml:base="[s3-bucket-uri]">
```
 
> 1. (Add meta data to document?) like so:
> 
> ```
> <item xmlns="http://libero.pub" xmlns:jats="http://jats.nlm.nih.gov">
>     <meta>
>         <id>article1</id>
>         <service>scholarly-articles</service>
>     </meta>
>     <jats:article></jats:article>
> </item>
> ```

`<service>` being the name of the content service where it'll be sent; `<id>` set from the value of `/article/front/article-meta/article-id[@pub-id-type="publisher-id"]` (set path ok for now, probably need to be configurable in the future).

> Should be the responsibility of the `content-store` to save it, so no need for storage on this side.

Yes, this is stateless.

**Comment by giorgiosironi** [comment]
Some infrastructure changes preparing for this new project:
- https://github.com/libero/scripts/pull/6
- https://github.com/libero/environments/pull/23
- https://github.com/libero/infrastructure/pull/14

**Comment by giorgiosironi** [comment]
The next question on infrastructure is how to provide secrets (the AWS credentials) to a container (what is likely this project could be deployed with). `docker-compose` has some support for secrets but only when used in swarm mode rather than on a single node. The lowest common denominator is to [add more configuration variables](https://github.com/libero/sample-configuration/blob/master/.env.dist) that will be passed in as environment variables only to the right containers; there may be a more specific way however.

**Comment by GiancarloFusiello** [comment]
Continuing from comments above to further define the service operations.

Assumptions:
- `.zip` files will be access from an AWS S3 bucket
- How `.zip` files are added to AWS S3 are not the concern of the ECA (eLife-style content adapter)

Configurations:
- AWS S3 bucket locations
- AWS S3 credentials
- (DB credentials?)
- Name of service to send PUT request to
- Address of service to send PUT request to

Process/processes:
1. Retrieve `.zip` file from AWS S3 `s3://unstable-elife-style-content-adapter-incoming`
2. Get article id from xpath `/article/front/article-meta/article-id[@pub-id-type="publisher-id"]`
3. Extract files and add to AWS S3 `s3://unstable-elife-style-content-adapter-expanded/<article-id>`
4. Wrap JATS content with an item tag like so: 
```xml
<item xmlns="http://libero.pub" xmlns:jats="http://jats.nlm.nih.gov">
  <jats:article></jats:article>
</item>
```
5. Add `xml:base=""` attribute to `jats:artcle` tag like so:
```xml
<jats:article xml:base="[s3-expanded-bucket-uri]">
```
6. (Add meta data to document?) like so:
```xml
<item xmlns="http://libero.pub" xmlns:jats="http://jats.nlm.nih.gov">
    <meta>
        <id>[article id]</id>
        <service>[service name from config file]</service>
    </meta>
    <jats:article></jats:article>
</item>
```
7. Make http PUT request to <address-from-config> with the resulting xml document as the body

@thewilkybarkid @giorgiosironi please can you edit/add missing information above?

**Comment by GiancarloFusiello** [comment]
Regarding the service being stateless, I think it depends on whether the different processes need to be queued, how the pipeline is initiated and how to retry a fail process.

To start the pipeline the service will either need to be notified of a zip file being added to an S3 bucket or periodically check a bucket for new zip files. If the service needs to poll the bucket then the service will need to keep a record of the bucket contents to see what has changed. Alternatively, the service could exposed an endpoint for incoming notifications to trigger the pipeline.

Options for receiving events from S3 are using SQS or writing a small Lambda function.

I'm also thinking that, as these processes become more complex, the pipeline shouldn't be a blocking process but a series of sequential processes.

How does all of this tie with the Dashboard?

**Comment by giorgiosironi** [comment]
(edited https://github.com/libero/libero/issues/131#issuecomment-483757252 to add all the relevant PRs for secrets management)

**Comment by giorgiosironi** [comment]
> Regarding the service being stateless, I think it depends on whether the different processes need to be queued, how the pipeline is initiated and how to retry a failed process.

I imagine that as soon as the code is executed in more than one process (rather than the result of a single HTTP POST) some state is necessary to carry over from one transaction to the next e.g. if the download of the zip doesn't happen in the handling of that POST notification from the bucket, there is some state in a database or internal queue that records a notification has happened and the next step still need to happen.

> To start the pipeline the service, it will either need to be notified of a zip file being added to an S3 bucket or periodically check a bucket for new zip files.

Periodically checking a bucket is not feasible because the only available operation is a full listing of the bucket returning keys in some order (alphabetical I think); for a small bucket like our own `s3://elife-production-final` (10K keys), this takes already 6-8 seconds.

> Alternatively, the service could expose an endpoint for incoming notifications to trigger the pipeline.

That's a good integration point

> Options for receiving events from S3 are using SQS or writing a Lambda function.

Choose your poison :cocktail: "exposing an endpoint" excludes SQS I think.

https://docs.aws.amazon.com/AmazonS3/latest/dev/NotificationHowTo.html#notification-how-to-event-types-and-destinations lists also SNS (SQS but with a topics rather than queues) as an option, which according to https://docs.aws.amazon.com/sns/latest/dg/sns-http-https-endpoint-as-subscriber.html may notify HTTP(S) destinations. Is that feasible?

> I'm also thinking that, as these processes become more complex, the pipeline shouldn't be a blocking process but a series of sequential processes.

If you mean don't block the HTTP response to the notification on executing all the activities, yes, likely.

> How does all of this tie with the Dashboard?

It doesn't, a Dashboard project doesn't exist yet so I suggest to ignore it.

**Comment by giorgiosironi** [comment]
Wrote to Ayappan from Exeter to find out more about the integration options for the article zip archive coming in. Probably getting some first answers tomorrow as there is some consultation with other people on their side.

**Comment by giorgiosironi** [comment]
Ayyappan:
<blockquote>
Dear Giorgio,

Below is the information I got it from the IJM development team'

1. Does Exeter directly integrate with an S3 bucket, with AWS credentials released from eLife? Or are there additional layers on top? We have various CloudGates integrations (external service) where I believe FTP/SFTP is used as a protocol from clients, and files end up in the S3 bucket as a result of the CloudGates translation.

<Exeter>For eLife we directly accessing the S3 bucket, and we haven't use any protocol. If you wish we will use FTP/SFTP.

2. What's the name of the software component on Exeter's side that performs this. Is Kriya the correct name for that?
<Exeter>Yes, the component name is Kriya

3. Would IJM use that same component as eLife?
<Exeter>Are you refer the Kriya as component?, then yes.

4. What other protocols are supported, if any? For example, HTTP POST or PUT with which Content-Type, Google Cloud Storage, Azure Storage, etc.
<Exeter>At present we are not using any protocol, but if you want then we will consider.

5. I know in Continuum videos are handled separately from Glencoe, and are not in the ingested .zip archive. Possibly this is because of their size? How would you deal with zips there are bigger than 100MB?
<Exeter>For eLife we are uploading the file upto 1GB (without video), so size will not make any problem.
</blockquote>

So S3 seems is the standard de facto here. I may be biased by previous experiences, so I suggest to scale the three notifications options (Lambda, SQS, SNS) from 1 to 3 on various concerns, sum the scores and choose the best:

| Concern | Lambda + custom POST | SQS queue | SNS + POST notification |
| --- | --- | --- | --- |
| Simplicity (inverse of number of lines of code required) |  |  |  | 
| Portability (inverse of lock in) |  |  |  | 
| Ease of applying updates |  |  |  | 
| Durability (notifications are not lost) | | | |

Feel free to add more rows and score this @GiancarloFusiello.

**Comment by giorgiosironi** [comment]
Also went through https://github.com/libero/libero/issues/131#issuecomment-483757252 reviews with @GiancarloFusiello yesterday and I have a few things to fix in that list.

**Comment by GiancarloFusiello** [comment]
>Feel free to add more rows and score this @GiancarloFusiello.

I'm not sure about how I would score each option as I have not implemented SQS or SNS previously.

As two options involve exposing an API to receive POST requests, I think we should start with this approach. Agreed?

**Comment by giorgiosironi** [comment]
> As two options involve exposing an API to receive POST requests, I think we should start with this approach. Agreed?

Agreed. The format of the request body may depend on which option, but it should be easy to change from one to the other if it's encapsulated in some function/module.

**Comment by giorgiosironi** [comment]
> I'm not sure about how I would score each option as I have not implemented SQS or SNS previously.

I haven't uses SNS or Lambda this way either, but scoring on a formula tries to be a [incremental improvement over going by gut feeling](https://www.businessinsider.com/daniel-kahneman-on-hiring-decisions-2013-1).

**Comment by GiancarloFusiello** [comment]
adding this to the list:

Configurations:
- AWS S3 bucket locations
- AWS S3 credentials
- (DB credentials?)
- Name of service to send PUT request to
- Address of service to send PUT request to

Process/processes:
1. Create endpoint to accept POST requests to trigger pipeline
2. Retrieve `.zip` file from AWS S3 `s3://unstable-elife-style-content-adapter-incoming`
3. Get article id from xpath `/article/front/article-meta/article-id[@pub-id-type="publisher-id"]`
4. Extract files and add to AWS S3 `s3://unstable-elife-style-content-adapter-expanded/<article-id>`
5. Wrap JATS content with an item tag like so: 
```xml
<item xmlns="http://libero.pub" xmlns:jats="http://jats.nlm.nih.gov">
  <jats:article></jats:article>
</item>
```
6. Add `xml:base=""` attribute to `jats:artcle` tag like so:
```xml
<jats:article xml:base="[s3://unstable-elife-style-content-adapter-expanded/<article-id>]">
```
7. (Add meta data to document?) like so:
```xml
<item xmlns="http://libero.pub" xmlns:jats="http://jats.nlm.nih.gov">
    <meta>
        <id>[article id]</id>
        <service>[service name from config file]</service>
    </meta>
    <jats:article></jats:article>
</item>
```
8. Make http PUT request to <address-from-config> with the resulting xml document as the body

@giorgiosironi @thewilkybarkid Does `6.` need to use `http`?

**Comment by giorgiosironi** [comment]
> Does 6. need to use http?

I think so, because otherwise `content-store` (receiving this XML) becomes dependent on S3. HTTPS over HTTP if it's free to implement (S3 buckets should support it out of the box with their own hostnames).

**Comment by GiancarloFusiello** [comment]
Do you know if the URL is different when using https?

**Comment by giorgiosironi** [comment]
Both of these work for me:
http://unstable-elife-style-content-adapter-expanded.s3.amazonaws.com/test.txt
https://unstable-elife-style-content-adapter-expanded.s3.amazonaws.com/test.txt

**Comment by GiancarloFusiello** [comment]
We still need to agree on which technology to use to implement this service.

Given that we need to: 
- expose an endpoint/endpoints
- run a series of tasks
- be able to re-run tasks on failure
- create application logs
- manage tasks state/history

We can either create code to do this or (following the walking skeleton) use Apache Airflow. Thoughts?

**Comment by giorgiosironi** [comment]
Using Airflow is reasonable for me, if you look at [random problems that come up daily from a production team](https://github.com/elifesciences/issues/issues/4889) it's clear this is an area where things can get complicated and need a robust architecture. I haven't seen other Libero projects built as a prototype to be replaced, but as production code; so if you think Airflow addresses these requirements (and possibly more that will come) go for it.

**Comment by thewilkybarkid** [comment]
:+1:

> I haven't seen other Libero projects built as a prototype to be replaced, but as production code; so if you think Airflow addresses these requirements (and possibly more that will come) go for it.

The scope of this ticket is to decide what to use; it if needs more experimentation then now's the time. (One of the tasks is 'Create ticket(s) for the implementation', so we won't be deploying anything here.)

**Comment by giorgiosironi** [comment]
The `walking-skeleton` has a [Airflow setup](https://github.com/libero/walking-skeleton/blob/master/docker-compose.yml#L101-L180), though it is residing inside the `article-store` (now `content-store`) to execute pipelines of tasks in there rather than working as an adapter.

**Comment by GiancarloFusiello** [comment]
I've had a look at this and the tasks it performs are not the same as outlined in this spike. Also it appears to be using a libero plugin that adds a custom python operator for integrating with rabbit mq. Airflow has official support for rabbit mq although, I've not looked into this in detail. Either way, using rabbit mq is out of scope currently.

I think it will be simpler if I create a new repo with an implementation of the processes that been outlined in this spike.

**Comment by giorgiosironi** [comment]
Yes, it's code from 1 year ago never used in production.

**Comment by GiancarloFusiello** [comment]
Added new repo: https://github.com/libero/elife-style-content-adapter-prototype

**Comment by GiancarloFusiello** [comment]
Looking at https://developer.mozilla.org/en-US/docs/Web/XML/xml:base It's recommended that we do not use `xml:base`, unless there is a transformation step I'm not aware of? 

Are figures/images considered as static content? If so, how is static content handled in Browser service? If we are using templating, could we do something like `<img src="{% static-url %}/figure-1.jpg"/>`?

Also, after discussing with Paul S, we agreed that the simplest approach for checking for new files in a S3 bucket is to use the Airflow [S3KeySensor](https://airflow.readthedocs.io/en/stable/_modules/airflow/sensors/s3_key_sensor.html). Basically, this polls a bucket at a given interval, which is not great, but would not require any additional resources other than a bucket meaning it would work on GCP also if required. Thoughts?

**Comment by thewilkybarkid** [comment]
> Looking at https://developer.mozilla.org/en-US/docs/Web/XML/xml:base It's recommended that we do not use `xml:base`, unless there is a transformation step I'm not aware of?

That's about HTML/browsers.
 
> Are figures/images considered as static content? If so, how is static content handled in Browser service? If we are using templating, could we do something like `<img src="{% static-url %}/figure-1.jpg"/>`?

Public URIs for images etc are handled by the Content Store (https://github.com/libero/content-store/pull/9), Browser will use them.

> Also, after discussing with Paul S, we agreed that the simplest approach for checking for new files in a S3 bucket is to use the Airflow [S3KeySensor](https://airflow.readthedocs.io/en/stable/_modules/airflow/sensors/s3_key_sensor.html). Basically, this polls a bucket at a given interval, which is not great, but would not require any additional resources other than a bucket meaning it would work on GCP also if required. Thoughts?

Sounds fine as a starting point at least. Also supporting things like GCP out of the box is useful, but shouldn't come at the expense of functionality (see also https://github.com/libero/content-store/pull/9#discussion_r282388967).

**Comment by GiancarloFusiello** [comment]
>That's about HTML/browsers.

Correct. How will `xml:base` be used in the resulting HTML? Given [libero/content-store#9 comment](https://github.com/libero/content-store/pull/9#discussion_r281533386) should `xml:base` be added?

>Public URIs for images etc are handled by the Content Store (libero/content-store#9), Browser will use them.

What does this service need to do for images in the xml?

**Comment by thewilkybarkid** [comment]
> Correct. How will `xml:base` be used in the resulting HTML?

It isn't. (It's the data used to generated the HTML, it's used if the `xlink:href` is not absolute.)

> What does this service need to do for images in the xml?

Make them publicly accessible (https://github.com/libero/libero/issues/132).

**Comment by GiancarloFusiello** [comment]
updating the list:

Configurations:
- AWS S3 bucket locations
- AWS S3 credentials
- DB credentials
- Name of service to send PUT request to
- Address of service to send PUT request to

Process/processes:
1. Implement Airflow's S3KeySensor and set poke interval to 30 minutes
2. Retrieve `.zip` file from AWS S3 `s3://unstable-elife-style-content-adapter-incoming`
3. Get article id from xpath `/article/front/article-meta/article-id[@pub-id-type="publisher-id"]`
4. Extract files and add to AWS S3 `s3://unstable-elife-style-content-adapter-expanded/<article-id>`
5. Wrap JATS content with an item tag like so: 
```xml
<item xmlns="http://libero.pub" xmlns:jats="http://jats.nlm.nih.gov">
  <jats:article></jats:article>
</item>
```
6. Add `xml:base=""` attribute to `jats:artcle` tag like so:
```xml
<jats:article xml:base="https://unstable-elife-style-content-adapter-expanded.s3.amazonaws.com/<article-id>">
```
7. Add meta data to document:
```xml
<item xmlns="http://libero.pub" xmlns:jats="http://jats.nlm.nih.gov">
    <meta>
        <id>[article id]</id>
        <service>[service name from config file]</service>
    </meta>
    <jats:article></jats:article>
</item>
```
8. Make http PUT request to <address-from-config> with the resulting xml document as the body

@giorgiosironi @thewilkybarkid comments or concerns?

**Comment by giorgiosironi** [comment]
> Implement Airflow's S3KeySensor

Looking at the source code, it will require a wildcard match since the key is not known beforehand (e.g. `article-*.zip` or similar

> and set poke interval to 30 minutes

That makes for a really long demo :thinking: any chance to set it lower?

**Comment by thewilkybarkid** [comment]
> > and set poke interval to 30 minutes
> 
> That makes for a really long demo 🤔 any chance to set it lower?

That's also not suitable for real usage.

**Comment by GiancarloFusiello** [comment]
>Looking at the source code, it will require a wildcard match since the key is not known beforehand (e.g. article-*.zip or similar

Correct. I was planning to use a wildcard to find all zip files and use the modified date to determine which files are considered new (or new since last run).

>That makes for a really long demo 🤔 any chance to set it lower?

Paul was the one to suggest 30 minutes. I was thinking more like once every 24 hours. I'm happy to make it as frequent as specified.

>That's also not suitable for real usage.

Do you have a suggestion?

**Comment by thewilkybarkid** [comment]
Items appearing is probably the result of someone pressing a button and expecting things to happening, so we’re probably talking seconds (for real usage).

**Comment by giorgiosironi** [comment]
Fair to say we are going to use Airflow, in-person discussion has been had.

Split:
- https://github.com/libero/libero/issues/154
- https://github.com/libero/libero/issues/155
- https://github.com/libero/libero/issues/156

Remaining tasks in this ticket:
- choose a name for the application
- external DAGs customization to be investigated (e.g. downloading from S3)

**Comment by GiancarloFusiello** [comment]
Regarding remote DAGs, currently it is not supported. Meaning that DAGs need to be present on the file system of all instances of Airflow.

However, there is this internal discussion about adding a DAG fetcher: https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-5+Remote+DAG+Fetcher
I'm quite confident I could write this for our use case. Thoughts on spending time doing this?

Other options are: 
- to add DAGs to a git repo and pull them into the project during deployment
- mount bucket as a volume using S3fuse (not recommended)
- download DAGs from bucket and mount directory as volume during deployment

**Comment by giorgiosironi** [comment]
> DAGs need to be present on the file system of all instances of Airflow.

This being a container, it seems this would be a read-only volume mounted on the various containers. The DAGs need to make it onto the host, I would leave open how they get there as it depends on the available infrastructure. In our unstable environment I agree it could easily be a Git repository checked out directly.

**Comment by GiancarloFusiello** [comment]
> This being a container, it seems this would be a read-only volume mounted on the various containers. 

Sounds like a good approach to me.

Now that we have determined that Airflow can be packaged/run independently of DAGs, I would like to separate Airflow and DAGs into independent repos. However, for the upcoming demo it may be easier to keep this all in one repo for now. Thoughts?

**Comment by GiancarloFusiello** [comment]
Closing this issue in favour of focused issues.
